### DISCUSSION TITLE
WebSockets/bringing HardLight to Leptos

### DISCUSSION THREAD

**Discussion by Zk2u** [leptos-rs/leptos#1512] [Q&A]
Hiya,

We've been looking into bringing [HardLight](https://github.com/valeralabs/hardlight) client support into WASM. It's currently built with Tokio, so we're thinking of shipping it with a stripped down version of Tokio which it heavily depends on. 

HL uses WebSockets to connect to an RPC server. The WebSocket is then used for both RPC requests and real-time server controlled events. Our plan is to use Leptos for our web app, and ship a HL client in it. 

I think Leptos is a good option for this because of its reactivity, which means (as long as it works) much of the real time functionality is awaiting new events from the Hardlight client over a Tokio channel, then writing to signals.

From your experience in Leptos, is this possible? If not, how would we make this possible? Looking forward to the discussions.

**Comment by gbj** [comment]
I'm not really sure what you're asking. Is the question: Is it possible to use websockets in the browser with Leptos? (Yes.) Is it a good idea to ship Tokio to the browser? (Probably not, but I guess depending on the circumstances--if initial loading time doesn't matter, then sure)

**Reply by Zk2u** [reply]
Both, I suppose. HardLight makes heavy use of Tokio's channels and green threads to work currently, so I'm guessing it would need a Tokio runtime running before it works. That includes a lot of event loops, which would make sense (if we were operating in pure JS) inside a web worker. We're just exploring if it's worth bringing it from native targets into wasm.

**Reply by gbj** [reply]
As I understand it, Tokio supports only a minimal set of its features in WASM, and I may be wrong but I think they actually mean `wasm32-wasi` (ie server-side WASM) not `wasm32-unknown-unknown` in the browser. It may well be the case that your library simply can't be compiled to WASM that can run in the browser at all. 

If it can be compiled to WASM that runs in the browser, then great!

https://docs.rs/tokio/latest/tokio/#supported-platforms
